File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel/Persistence/CdmFolder/DocumentPersistence.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

namespace Microsoft.CommonDataModel.ObjectModel.Persistence.CdmFolder
{
    using Microsoft.CommonDataModel.ObjectModel.Cdm;
    using Microsoft.CommonDataModel.ObjectModel.Enums;
    using Microsoft.CommonDataModel.ObjectModel.Persistence.CdmFolder.Types;
    using Microsoft.CommonDataModel.ObjectModel.Utilities;
    using Microsoft.CommonDataModel.ObjectModel.Utilities.Logging;
    using Newtonsoft.Json;
    using System.Linq;

    public class DocumentPersistence
    {
        private static readonly string Tag = nameof(DocumentPersistence);

        /// <summary>
        /// Whether this persistence class has async methods.
        /// </summary>
        public static readonly bool IsPersistenceAsync = false;

        /// <summary>
        /// The file format/extension types this persistence class supports.
        /// </summary>
        public static readonly string[] Formats = { PersistenceLayer.CdmExtension };

        /// <summary>
        /// The maximum json semantic version supported by this ObjectModel version.
        /// </summary>
        public static readonly string JsonSemanticVersion = CdmDocumentDefinition.CurrentJsonSchemaSemanticVersion;

        public static CdmDocumentDefinition FromObject(CdmCorpusContext ctx, string name, string nameSpace, string path, DocumentContent obj)
        {
            var doc = ctx.Corpus.MakeObject<CdmDocumentDefinition>(CdmObjectType.DocumentDef, name);
            doc.FolderPath = path;
            doc.Namespace = nameSpace;

            if (!StringUtils.IsBlankByCdmStandard(obj.Schema))
            {
                doc.Schema = obj.Schema;
            }

            if (!StringUtils.IsBlankByCdmStandard(obj.DocumentVersion))
            {
                doc.DocumentVersion = obj.DocumentVersion;
            }

            if (obj.Imports != null)
            {
                foreach (var importObj in obj.Imports)
                {
                    doc.Imports.Add(ImportPersistence.FromData(ctx, importObj));
                }
            }

            if (obj.Definitions != null)
            {
                for (int i = 0; i < obj.Definitions.Count; i++)
                {
                    dynamic d = obj.Definitions[i];
                    if (d["dataTypeName"] != null)
                    {
                        doc.Definitions.Add(DataTypePersistence.FromData(ctx, d));
                    }
                    else if (d["purposeName"] != null)
                    {
                        doc.Definitions.Add(PurposePersistence.FromData(ctx, d));
                    }
                    else if (d["attributeGroupName"] != null)
                    {
                        doc.Definitions.Add(AttributeGroupPersistence.FromData(ctx, d));
                    }
                    else if (d["traitName"] != null)
                    {
                        doc.Definitions.Add(TraitPersistence.FromData(ctx, d));
                    }
                    else if (d["traitGroupName"] != null)
                    {
                        doc.Definitions.Add(TraitGroupPersistence.FromData(ctx, d));
                    }
                    else if (d["entityShape"] != null)
                    {
                        doc.Definitions.Add(ConstantEntityPersistence.FromData(ctx, d));
                    }
                    else if (d["entityName"] != null)
                    {
                        doc.Definitions.Add(EntityPersistence.FromData(ctx, d));
                    }
                }
            }

            var isResolvedDoc = false;
            if (doc.Definitions.Count == 1 && doc.Definitions[0] is CdmEntityDefinition entity)
            {
                var resolvedTrait = entity.ExhibitsTraits.Item("has.entitySchemaAbstractionLevel");
                // Tries to figure out if the document is in resolved form by looking for the schema abstraction trait
                // or the presence of the attribute context.
                isResolvedDoc = resolvedTrait != null && string.Equals((resolvedTrait as CdmTraitReference).Arguments[0].Value, "resolved");
                isResolvedDoc = isResolvedDoc || entity.AttributeContext != null;
            }

            if (!StringUtils.IsBlankByCdmStandard(obj.JsonSchemaSemanticVersion))
            {
                doc.JsonSchemaSemanticVersion = obj.JsonSchemaSemanticVersion;
                if (CompareJsonSemanticVersion(ctx, doc.JsonSchemaSemanticVersion) > 0)
                {
                    if (isResolvedDoc)
                    {
                        Logger.Warning(ctx, Tag, nameof(FromObject), null, CdmLogCode.WarnPersistUnsupportedJsonSemVer, JsonSemanticVersion, doc.JsonSchemaSemanticVersion);
                    }
                    else
                    {
                        Logger.Error(ctx, Tag, nameof(FromObject), null, CdmLogCode.ErrPersistUnsupportedJsonSemVer, JsonSemanticVersion, doc.JsonSchemaSemanticVersion);
                    }
                }
            }
            else
            {
                Logger.Warning(ctx, Tag, nameof(FromObject), null, CdmLogCode.WarnPersistJsonSemVerMandatory);
            }

            return doc;
        }

        public static CdmDocumentDefinition FromData(CdmCorpusContext ctx, string docName, string jsonData, CdmFolderDefinition folder)
        {
            var obj = JsonConvert.DeserializeObject<DocumentContent>(jsonData, PersistenceLayer.SerializerSettings);
            return FromObject(ctx, docName, folder.Namespace, folder.FolderPath, obj);
        }

        public static DocumentContent ToData(CdmDocumentDefinition instance, ResolveOptions resOpt, CopyOptions options)
        {
            return new DocumentContent
            {
                Schema = instance.Schema,
                JsonSchemaSemanticVersion = instance.JsonSchemaSemanticVersion,
                Imports = Utils.ListCopyData<Import>(resOpt, instance.Imports, options),
                Definitions = CopyDataUtils.ListCopyData(resOpt, instance.Definitions, options),
                DocumentVersion = instance.DocumentVersion
            };
        }

        /// <summary>
        /// Compares the document version with the json semantic version supported.
        /// </summary>
        /// <param name="documentSemanticVersion"></param>
        /// <returns>
        /// 1 => if documentSemanticVersion > JsonSemanticVersion
        /// 0 => if documentSemanticVersion == JsonSemanticVersion or if documentSemanticVersion is invalid
        /// -1 => if documentSemanticVersion < JsonSemanticVersion
        /// </returns>
        private static int CompareJsonSemanticVersion(CdmCorpusContext ctx, string documentSemanticVersion)
        {
            var docSemanticVersionSplit = documentSemanticVersion.Split('.');
            var currSemanticVersionSplit = JsonSemanticVersion.Split('.').Select(value => int.Parse(value)).ToList();

            if (docSemanticVersionSplit.Length != 3)
            {
                Logger.Warning(ctx, Tag, nameof(CompareJsonSemanticVersion), null, CdmLogCode.WarnPersistJsonSemVerInvalidFormat);
                return 0;
            }

            for (var i = 0; i < 3; ++i)
            {
                if (!int.TryParse(docSemanticVersionSplit[i], out int version))
                {
                    Logger.Warning(ctx, Tag, nameof(CompareJsonSemanticVersion), null, CdmLogCode.WarnPersistJsonSemVerInvalidFormat);
                    return 0;
                }

                if (version != currSemanticVersionSplit[i])
                {
                    return version < currSemanticVersionSplit[i] ? -1 : 1;
                }
            }

            return 0;
        }
    }
}

File: objectModel/TypeScript/Persistence/CdmFolder/DocumentPersistence.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import { CdmFolder } from '..';
import {
    CdmConstants,
    CdmCorpusContext,
    CdmDocumentDefinition,
    CdmEntityDefinition,
    CdmFolderDefinition,
    cdmObjectType,
    CdmTraitReference,
    copyOptions,
    Logger,
    cdmLogCode,
    resolveOptions,
    CdmTraitReferenceBase,
    StringUtils
} from '../../internal';
import * as copyDataUtils from '../../Utilities/CopyDataUtils';
import {
    AttributeGroup,
    ConstantEntity,
    DataType,
    DocumentContent,
    Entity,
    Import,
    Purpose,
    Trait
} from './types';

export class DocumentPersistence {
    private static TAG: string = DocumentPersistence.name;

    /**
     * Whether this persistence class has async methods.
     */
    public static readonly isPersistenceAsync: boolean = false;

    /**
     * The file format/extension types this persistence class supports.
     */
    public static readonly formats: string[] = [CdmConstants.cdmExtension];

    public static fromObject(ctx: CdmCorpusContext, name: string, namespace: string, path: string, object: DocumentContent): CdmDocumentDefinition {
        const document: CdmDocumentDefinition = ctx.corpus.MakeObject(cdmObjectType.documentDef, name);
        document.folderPath = path;
        document.namespace = namespace;

        if (object) {
            if (!StringUtils.isBlankByCdmStandard(object.$schema)) {
                document.schema = object.$schema;
            }
            // support old model syntax
            if (!StringUtils.isBlankByCdmStandard(object.schemaVersion)) {
                document.jsonSchemaSemanticVersion = object.schemaVersion;
            }

            if (!StringUtils.isBlankByCdmStandard(object.documentVersion)) {
                document.documentVersion = object.documentVersion;
            }

            if (object.imports) {
                for (const importObj of object.imports) {
                    document.imports.push(CdmFolder.ImportPersistence.fromData(ctx, importObj));
                }
            }
            if (object.definitions && Array.isArray(object.definitions)) {
                for (const definition of object.definitions) {
                    if ('dataTypeName' in definition) {
                        document.definitions.push(CdmFolder.DataTypePersistence.fromData(ctx, definition));
                    } else if ('purposeName' in definition) {
                        document.definitions.push(CdmFolder.PurposePersistence.fromData(ctx, definition));
                    } else if ('attributeGroupName' in definition) {
                        document.definitions.push(CdmFolder.AttributeGroupPersistence.fromData(ctx, definition));
                    } else if ('traitName' in definition) {
                        document.definitions.push(CdmFolder.TraitPersistence.fromData(ctx, definition));
                    } else if ('traitGroupName' in definition) {
                        document.definitions.push(CdmFolder.TraitGroupPersistence.fromData(ctx, definition));
                    } else if ('entityShape' in definition) {
                        document.definitions.push(CdmFolder.ConstantEntityPersistence.fromData(ctx, definition));
                    } else if ('entityName' in definition) {
                        document.definitions.push(CdmFolder.EntityPersistence.fromData(ctx, definition));
                    }
                }
            }
        }

        let isResolvedDoc: boolean = false;
        if (document.definitions.length === 1 && document.definitions.allItems[0].objectType == cdmObjectType.entityDef) {
            const entity: CdmEntityDefinition = document.definitions.allItems[0] as CdmEntityDefinition;
            const resolvedTrait: CdmTraitReferenceBase = entity.exhibitsTraits.item('has.entitySchemaAbstractionLevel');
            // Tries to figure out if the document is in resolved form by looking for the schema abstraction trait
            // or the presence of the attribute context.
            isResolvedDoc = resolvedTrait != null && (resolvedTrait as CdmTraitReference).arguments.allItems[0].value == 'resolved';
            isResolvedDoc = isResolvedDoc || !!entity.attributeContext;
        }

        if (!StringUtils.isBlankByCdmStandard(object.jsonSchemaSemanticVersion)) {
            document.jsonSchemaSemanticVersion = object.jsonSchemaSemanticVersion;
            if (DocumentPersistence.compareJsonSemanticVersion(ctx, document.jsonSchemaSemanticVersion) > 0) {
                if (isResolvedDoc) {
                    Logger.warning(ctx, this.TAG, this.fromObject.name, null, cdmLogCode.WarnPersistUnsupportedJsonSemVer, CdmDocumentDefinition.currentJsonSchemaSemanticVersion, document.jsonSchemaSemanticVersion);
                } else {
                    Logger.error(ctx, this.TAG, this.fromObject.name, null, cdmLogCode.ErrPersistUnsupportedJsonSemVer, CdmDocumentDefinition.currentJsonSchemaSemanticVersion, document.jsonSchemaSemanticVersion);
                }
            }
        } else {
            Logger.warning(ctx, this.TAG, this.fromObject.name, null, cdmLogCode.WarnPersistJsonSemVerMandatory);
        }

        return document;
    }

    public static fromData(ctx: CdmCorpusContext, docName: string, jsonData: string, folder: CdmFolderDefinition): CdmDocumentDefinition {
        const obj = JSON.parse(jsonData);

        return DocumentPersistence.fromObject(ctx, docName, folder.namespace, folder.folderPath, obj);
    }

    public static toData(instance: CdmDocumentDefinition, resOpt: resolveOptions, options: copyOptions): DocumentContent {
        return {
            $schema: instance.schema,
            jsonSchemaSemanticVersion: instance.jsonSchemaSemanticVersion,
            imports: copyDataUtils.arrayCopyData<Import>(resOpt, instance.imports, options),
            definitions: copyDataUtils.arrayCopyData<Trait | DataType | Purpose | AttributeGroup | Entity | ConstantEntity>(
                resOpt, instance.definitions, options),
            documentVersion: instance.documentVersion
        };
    }

    /**
     * Compares the document version with the json semantic version supported.
     * 1 => if documentSemanticVersion > jsonSemanticVersion
     * 0 => if documentSemanticVersion == jsonSemanticVersion or if documentSemanticVersion is invalid
     * -1 => if documentSemanticVersion < jsonSemanticVersion
     */
    private static compareJsonSemanticVersion(ctx: CdmCorpusContext, documentSemanticVersion: string): number {
        const docSemanticVersionSplit: number[] = documentSemanticVersion.split(".").map(x => Number(x));
        const currSemanticVersionSplit: number[] = CdmDocumentDefinition.currentJsonSchemaSemanticVersion.split(".").map(x => Number(x));

        if (docSemanticVersionSplit.length !== 3 || docSemanticVersionSplit.includes(NaN)) {
            Logger.warning(ctx, this.TAG, this.compareJsonSemanticVersion.name, null, cdmLogCode.WarnPersistJsonSemVerInvalidFormat);
            return 0;
        }

        for (let i = 0; i < 3; ++i) {
            if (docSemanticVersionSplit[i] !== currSemanticVersionSplit[i]) {
                return docSemanticVersionSplit[i] < currSemanticVersionSplit[i] ? -1 : 1;
            }
        }

        return 0;
    }
}
